### ISSUE TITLE
Make indicies stepping more effecient

### ISSUE THREAD

**Issue by tannergooding** [dotnet/runtime#28389]
_From @ericstj on October 20, 2017 16:19_

In many cases we do operations using only the linearized index.   There are cases where we instead need to use an indices array/span.  In these cases we should step the indices array rather than continuous composing/decomposing indices from a linearized index.

_Copied from original issue: dotnet/corefxlab#1867_

**Comment by deeprobin** [comment]
@ericstj Does this performance problem still exist?
If so, could you provide a small code example?

**Comment by ericstj** [comment]
It's been a while since I've written that but I'm sure it still exists since we haven't done much here.  Looking at the history I was writing API docs at that time.  TL;DR Tensor access has a lot of perf problems today so this is probably more of a rethink/redesign problem than an up-for-grabs issue.

Tensors provide multi-dimensional array access which the JIT knows nothing about, so if you try to step through one of the dimensions the JIT cannot see if the stepping is actually just stepping by an element size over a pointer.  So what ends up happening is integer stepping of an index, then multiplication and addition to determine the pointer offset, and bounds checks on those.  We also have multiple virtual calls in the form of the indexer and the linearized GetValue and SetValue which don't help.  So lots multiplication, jumps, and branches.

Some operators in TensorArithmetic try to avoid multi-dimensional access and have DenseTensor specific implementations to get better performance.  They get spans of both left hand side and right hand side and try to work in terms of linear offset.  In this way they can stay in the realm of pointer-adds without bound checks (no virtual calls, multiplication, little branching).

@michaelgsharp and I just looked at a similar problem.  We fixed it by lifting the indexing logic (calculation offset using of strides and indices) up to the app level.  This is essentially what everyone who does large amounts of element access needs to do today: understand how to translate that work to accesses on a single dimensional span to cut out multiplication and virtual calls.  We need to figure out how to make this easier for folks to do -- either by teaching the JIT how to understand our multi-dimensional access or by making it easier for folks 

I think this issue was probably talking about the cases where we were using a lineralized routine, but translate the access to indices access.  Here's a case like this (TensorArithmetic was part of the library, not tests at the time):
https://github.com/dotnet/runtime/blob/8ced105bdf70ca566df60b7894de44d28271e09e/src/libraries/System.Numerics.Tensors/tests/TensorArithmetic.cs#L667-L671
So that's stepping a linear index, but then translating that to indices, which then may translate it back based on the Tensor implementation.
